add guidance on implementing identity&auth workflow (#2906)
awslabs/smithy: docs/source-2.0/guides/client-guidance/identity-auth/index.md

# Identity and Authentication

Operations supported by a Smithy service generally require a mechanism for
authenticating the client's identity. This section describes how to model the
workflow of identity retrieval and request authentication in Smithy clients.

## Identity

An Identity is an entity representing **who the caller is**. Abstractly, the
caller's identity could be anonymous, a token, a public/private key, etc.

Identity types in the client runtime should typically extend from a base

> interface that exposes a single value: the expiration time of the identity. The

Author: I think the only real need here is a base interface to reference in the other interfaces, expiration makes sense since basically every identity will have that but we don't need to state it in this imperative way.

API should then branch per auth scheme. Sometimes an Identity type may be
re-used across multiple schemes.

```java
public interface Identity {
	Instant expiration();
}

// for smithy.api#bearerTokenAuth
public interface BearerTokenIdentity extends Identity {
	String token();
}

// for aws.auth#sigv4, aws.auth#sigv4a
public interface SigV4CredentialIdentity extends Identity {
	String accessKeyId();
	String secretAccessKey();
	String sessionToken();
}
```

## IdentityResolver

An IdentityResolver implements the retrieval of a single type of identity.

The IdentityResolver interface should be modeled as follows:

```java
public interface IdentityResolver<TIdentity extends Identity> {

>     TIdentity resolve(Object properties);

Author: The identity resolver / signer property bags are not formally defined here so I've just left them with an `Object` placeholder. I think the typed property bag construct needs its own section in these guidance docs.

the service.

```java
public class MyServiceClientConfig {
    private final IdentityResolver<SigV4CredentialIdentity> sigV4CredentialResolver;
    private final IdentityResolver<BearerTokenIdentity> bearerTokenResolver;

    // other configuration...
}
```

## Signer

A Signer is an entity representing **a way to generate a signature for a
request**. A signature is metadata attached to a request that will be sent to
the service in order to allow the service to authenticate the Smithy client
caller's identity.

For example, an operation that uses a bearer token for
identification would "sign" an outgoing HTTP request by presenting the value of
the token in the Authorization header.

All signers require, at a minimum, a request to sign and an identity with which
to sign. A Signer should be modeled like so:

```java
// This example interface modifies the transport message in-place with the
// signature. Depending on the surrounding runtime, the caller may instead wish to
// model the Signer as an interface that returns a copied, modified transport
// message instead.
public interface Signer<TIdentity extends Identity, TMessage extends Message> {

>     void sign(TIdentity identity, TMessage message, Object properties);

Author: The original doc assumed HTTP, I know formally we want to speak in terms of transport agnosticism so I've parameterized the transport type.

}
```

## AuthScheme

An AuthScheme self-describes a single flow (combination of an identity resolver
and signer) through which a Smithy client authenticates a request:

```java
public interface AuthScheme<TIdentity extends Identity, TMessage extends Message> {
    /**
     * Unique identifier for this auth scheme. Typically corresponds to the ID of a
     * Smithy IDL auth trait, e.g. "smithy.api#httpBearerAuth".
     */
	String schemeId();

    /**
     * Provides an Identity Resolver for this authentication scheme.

>      * This API can return a nullish value indicating that an identity

Reviewer: ```suggestion
     * This API can return a null value indicating that an identity
```

     * resolver of this scheme's type is not available in the current client
     * environment.
     */
    IdentityResolver<TIdentity> identityResolver(IdentityResolverConfig config);

    /**
     * Provides a Signer for this authentication scheme from an unspecified
     * source. An actual implementation might source the Signer from client
     * configuration, or perhaps provide its own implementation directly.
     */
    Signer<TIdentity, TMessage> signer();
}

public interface IdentityResolverConfig {
    IdentityResolver<?> getIdentityResolver(String schemeId);
}
```

A list of supported AuthSchemes should be configurable by the end-user of the
Smithy client. The default "constructor" for a Smithy client should typically

> pre-load a list of supported AuthSchemes on client configuration.

Reviewer: ```suggestion
pre-load a list of supported `AuthScheme`s on client configuration.
```

```java
// in this example, the service supports some combination of
// smithy.api#httpBearerAuth and aws.auth#sigv4

> public MyServiceClientConfig defaultConfig() {

Author: These types of code snippets seem useful to the reader but as we build out these docs we want to make sure we have a standard for stuff like this (i.e. showing how to bootstrap a default config).

}
```

### AuthSchemeOption

An AuthSchemeOption represents a **possible** AuthScheme that the current
request may use:

``` java
public interface AuthSchemeOption {
    /**
     * The unique ID of the scheme to use.
     */
    String schemeId();

    /**
     * Opaque container for additional context to be passed to the scheme's
     * IdentityResolver when retrieving Identity.
     */
    Object identityProperties();

    /**
     * Opaque container for additional context to be passed to the scheme's
     * Signer when signing the request. These allow the Signer to use
     * information from the request, client, etc. in the signature generation process.
     *
     * For example, a SigV4-enabled service would encode the "region" parameter
     * into its signer properties.
     */
    Object signerProperties();
}
```

After calling into the AuthSchemeResolver, the Smithy client should choose the
first AuthScheme in the resulting list of options that it supports. A scheme is
considered "supported" if the client is configured with a scheme with its ID
**and** if the identity resolver for the scheme is available.

This may be implemented like the following:

```java

> public void resolveAuthScheme(OperationContext ctx) {

Author: Again, including snippets like this seems helpful but we should be consistent. I elected to just model the examples as using this opaque / undescribed "operation context" input that has the stuff you'd expect on a request.

        }

        return new SelectedAuthScheme(
            found.get(),
            option.identityProperties(),
            option.signerProperties()
        );
    }

    return null;
}

record SelectedAuthScheme(
    AuthScheme<?, ?> scheme,
    Object identityProperties,
    Object signerProperties
) {}
```

## Order of Operations

The Smithy client should conduct the auth flow in an operation call as follows:

1. Auth scheme resolution (AuthSchemeResolver::resolveAuthScheme) is called
   **within** the retry loop. The selected auth scheme and identity / signer
   properties are stored in operation context for future use.
1. (Endpoint resolution)[TODO: NEED ANCHOR] is called. The signer properties
   from the resolved endpoint are **merged** into the ones sourced from auth
   scheme resolution.
1. Retrieve the IdentityResolver from the previously-resolved AuthScheme
   (AuthScheme::identityResolver). Identity resolution
   (IdentityResolver::resolve) is called with the identity properties sourced from
   scheme resolution.
1. Retrieve the Signer from the previously-resolved AuthScheme
   (AuthScheme::signer). Request signing (Signer::sign) is called with the

>    merged signer properties from scheme and endpoint resolution.

Reviewer: These both mix passive and imperative, a consistent tense should be used

## FAQ

> ### Wow, this seems like a lot. Do I really need all of these abstractions just to decide how to set an Authorization header?

Author: Probably controversial but this felt absolutely worth stating IMO. I think a lot of client designers _will_ sort of stick to baking in a single transport type (probably HTTP). The language of the question is probably too colloquial ofc.

The example set of interfaces provided in this listing represents the most
rigorous possible solution for implementing Identity & Auth. These interfaces
support multiple authentication schemes, _across_ multiple identity types and
transport message types. The Smithy client implementor is free to simplify

> along any of those dimensions as fits their service needs.

Reviewer: ```suggestion
along any of those dimensions as fits their needs.
```
